proposals/p2665.md

```diff
@@ -0,0 +1,209 @@
+# Semicolons terminate statements
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/2665)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Background](#background)
+    -   [Discussion in Carbon](#discussion-in-carbon)
+    -   [In other languages](#in-other-languages)
+        -   [Requiring semicolons](#requiring-semicolons)
+        -   [Optional semicolons](#optional-semicolons)
+-   [Proposal](#proposal)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Optional semicolons](#optional-semicolons-1)
+
+<!-- tocstop -->
+
+## Abstract
+
+Statements, declarations, and definitions will terminate with either a semicolon
+(`;`) or a close curly brace (`}`). Semicolons are never optional.
+
+For example, with a semicolon, `x = x + 2;` or `class C;`. With a close curly
+brace, `for ( ... ) { ... }`, or `class C { ...}`.
+
+This does not affect any approved proposal; rather, it makes an important
+assumption explicit.
+
+## Problem
+
+Statements need some system for separation. There are two main options for this:
+
+1. Require semicolons to terminate statements.
+2. Automatically determine where statements terminate.
+    - Some languages, such as Python, define a syntax where a newline terminates
+      statements.
+    - Other languages, such as Javascript, require semicolons but define rules
+      for semicolon insertion.
+
+Although Carbon's design currently assumes semicolons are required, it hasn't
+been directly addressed by a proposal.
+
+## Background
+
+### Discussion in Carbon
+
+This was discussed on leads issue
+[#1924: Semicolon](https://github.com/carbon-language/carbon-lang/issues/1924).
+Some rationale is provided there, stemming from discussion
+[#1739: Semicolon](https://github.com/carbon-language/carbon-lang/discussions/1739).
+
+### In other languages
+
+#### Requiring semicolons
+
+In C++, C#, and Java, semicolons are always required.
+
+In Rust, semicolons are generally required, but may be omitted for an
+[implicit return](https://doc.rust-lang.org/std/keyword.return.html).
+
+#### Optional semicolons
+
+In Python, a line is a
+[simple statement](https://docs.python.org/3/reference/simple_stmts.html), and
+parentheses are an idiomatic way to create multi-line statements. Semicolons may
+be used to explicitly separate statements. For example:
+
+```python
+value = (
+  "text"
+)
+a = 1; b = 2; c = 3
+```
+
+Swift takes a different approach, where
+[statement separation is part of the grammar](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/statements).
```

Comment:
I think your first example gives errors:
https://swift.compiler-explorer.com/z/E9EPs71Ez

The linked code is an error, but I think it's distinct for the empty tuple since `make()` is a function call and `()` has no value being discarded. I've added documentation of this as an example of ambiguity.

I would tend to agree that Swift and Kotlin are closer to JavaScript than Python, but I'm wary of the JavaScript comparison for ASI advice. I think there's an important distinction that Swift has better handling of potential mistakes, as demonstrated above, which I think is also at the root of why JavaScript discourages relying on ASI -- it doesn't handle mistakes well.

I've adjusted a little here, but is there more you're looking for, given your comparison?

Reply:
I've tried to document the poor documentation of Swift, but I do want to address one detail in your reply:

> By "statement separation is part of the grammar", I think you mean "the semicolon is simply optional in the grammar, and isn't needed to parse Swift properly", which is the claim that I've seen made in other forums, and that's not right

My perspective from flipping through semicolon approaches is there are two fundamental approaches:

1. The language requires semicolons, and there's a way to insert them, as in JavaScript or Go.
2. The language has optional semicolons, and appropriate statement termination is grammatically determined, as in Python, Swift, or Kotlin.

I understand that these two approaches may yield similar results, but I think it's an interesting difference in philosophies. I think Swift is closer to the second. Go and JavaScript are explicitly in the first category. That's what I mean when I say statement separation is part of the grammar.

Note, "properly" sounds subjective to me. While I did mean something different, I actually would be willing to agree the semicolon is optional other than for multi-statement lines. This may just be a different perspective -- while errors can occur with particular code structures, that doesn't mean Swift is wrong for how it treats those code structures.